iib/workers/dogpile_cache.py

```diff
@@ -0,0 +1,75 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+import os
+import threading
+
+from celery import current_task
+from dogpile.cache import make_region
+
+from iib.workers.config import get_worker_config
+
+
+def generate_cache_key(fname, *args, **kwargs):
+    """Generate key that is used in dogpile cache."""
+    namespace = dogpile_namespace.get_dogpile_namespace()
+    arguments = "_".join(
+        [str(arg) for arg in args] + [f"{kwarg}={kwargs[kwarg]}" for kwarg in kwargs]
+    )
+    return f"{namespace}_{fname}_{arguments}"
+
+
+def create_dogpile_region():
+    """Load setting from config and creates dogpile region."""
+    conf = get_worker_config()
+
+    return make_region().configure(
+        conf.iib_dogpile_backend,
+        expiration_time=conf.iib_dogpile_expiration_time,
+        arguments=conf.iib_dogpile_arguments,
+    )
+
+
+class DogpileNamespace:
+    """Used to prevent overwriting variable when running in parallel."""
+
+    def __init__(self):
+        """Initialize dogpile namespace."""
+        self.ns = {}
+
+    def get_dogpile_namespace(self):
+        """Get dogpile namespace."""
+        try:
+            thread_id = threading.current_thread().native_id
+        except AttributeError:
+            thread_id = ''
+
+        try:
+            celery_id = current_task.request.id
+        except AttributeError:
+            celery_id = ''
+
+        return self.ns.get(os.getpid(), {}).get(thread_id, {}).get(celery_id, '')
+
+    def set_dogpile_namespace(self, build_id):
+        """Set dogpile namespace."""
+        pid = os.getpid()
+        if pid not in self.ns:
+            self.ns[os.getpid()] = {}
+
+        try:
+            thread_id = threading.current_thread().native_id
+        except AttributeError:
+            thread_id = ''
+        if thread_id not in self.ns[pid]:
+            self.ns[pid][thread_id] = {}
+
+        try:
+            celery_id = current_task.request.id
+        except AttributeError:
+            celery_id = ''
+        if current_task.request.id not in self.ns[pid][thread_id]:
+            self.ns[pid][thread_id][celery_id] = ''
+
+        self.ns[pid][thread_id][celery_id] = str(build_id)
+
+
+dogpile_namespace = DogpileNamespace()
```

Comment:
Could you please define this at the top of the file?

Reply:
I don't think it's possible since the class is defined in the file and Python source is read starting at the top.

Reply:
Yeah, Matt is right.